### DISCUSSION TITLE
How to refactor configuration properties?

### DISCUSSION THREAD

**Discussion by tobiasschaefer** [micronaut-projects/micronaut-core#4958] [Q&A] [answered]
How can configuration properties be refactored while still supporting the old structure?

E.g. I have a supported property "x.y..z" and I want to remove "y" so that it is "x.z" but I still want to support x.y.z and preferably log a warning if it is used?

This is the existing structure:

```java
@ConfigurationProperties( "x")
public interface Configuration {

    @NotNull
    Y getY();

    @ConfigurationProperties("y")
    interface Y {

        @Bindable(defaultValue = "false")
        boolean isZ();
    }
}
```

This would be the new structure but it misses "y" completely:

```java
@ConfigurationProperties( "x")
public interface Configuration {

    @Bindable(defaultValue = "false")
     boolean isZ();
}
```

So how must I implement to support both?

Do you have any hints?

**Comment by graemerocher** [comment, marked as answer]
Don't think this is currently supported with interfaces, but I recommend reporting an issue with a feature request and we can see what we can do

**Reply by tobiasschaefer** [reply]
Actually, my use case is very specific. I guess it would need a more generic concept how to handle deprecated properties.

@graemerocher: Shall I file an issue for that instead? Maybe simply log a warning to every configuration (node and leaf)  that as been marked as `@Deprecated`.

Note: I'm now checking `environment.getPropertyEntries("x.y").isEmpty()` and throwing an exception on startup if there are any old configurations. That is sufficient for my use case.

**Reply by graemerocher** [reply]
Yes file an issue

**Reply by tobiasschaefer** [reply]
Files as https://github.com/micronaut-projects/micronaut-core/issues/4967
